In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

data = pd.read_csv('standardized_output.csv')

# Eğitim ve test setlerine ayırma (%80 eğitim, %20 test)
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)

C:\Users\ecegu\anaconda3\lib\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
C:\Users\ecegu\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.4' currently installed).
  from pandas.core import (


###  Linear Regression

In [2]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

model = LinearRegression()
model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])

# Tahminler yapma
y_pred = model.predict(test_data.drop(columns=['BAD']))

# Performans değerlendirmesi
print("Linear Regression MSE:", mean_squared_error(test_data['BAD'], y_pred))
print("Linear Regression MAE:", mean_absolute_error(test_data['BAD'], y_pred))
print("Linear Regression R2:", r2_score(test_data['BAD'], y_pred))

Linear Regression MSE: 0.06085011605395081
Linear Regression MAE: 0.13615216766640215
Linear Regression R2: 0.20233901903797502


### Linear Regression Hiperparametre Optimizasyonu (Ridge ve Lasso)

In [3]:
from sklearn.linear_model import Ridge, Lasso

# Ridge Regression
ridge_model = Ridge(alpha=1.0)
ridge_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_ridge = ridge_model.predict(test_data.drop(columns=['BAD']))

print("Ridge Regression MSE:", mean_squared_error(test_data['BAD'], y_pred_ridge))
print("Ridge Regression MAE:", mean_absolute_error(test_data['BAD'], y_pred_ridge))
print("Ridge Regression R2:", r2_score(test_data['BAD'], y_pred_ridge))

# Lasso Regression
lasso_model = Lasso(alpha=0.1)
lasso_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_lasso = lasso_model.predict(test_data.drop(columns=['BAD']))

print("Lasso Regression MSE:", mean_squared_error(test_data['BAD'], y_pred_lasso))
print("Lasso Regression MAE:", mean_absolute_error(test_data['BAD'], y_pred_lasso))
print("Lasso Regression R2:", r2_score(test_data['BAD'], y_pred_lasso))

Ridge Regression MSE: 0.06085077771025895
Ridge Regression MAE: 0.136149066649526
Ridge Regression R2: 0.20233034563759333
Lasso Regression MSE: 0.07634138506339005
Lasso Regression MAE: 0.15879247483356274
Lasso Regression R2: -0.0007301225797695654


### Random Forest

In [4]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest modeli oluşturma ve eğitme
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_rf = rf_model.predict(test_data.drop(columns=['BAD']))

# Performans değerlendirmesi
print("Random Forest MSE:", mean_squared_error(test_data['BAD'], y_pred_rf))
print("Random Forest MAE:", mean_absolute_error(test_data['BAD'], y_pred_rf))
print("Random Forest R2:", r2_score(test_data['BAD'], y_pred_rf))

Random Forest MSE: 0.030506092124814265
Random Forest MAE: 0.06962852897473996
Random Forest R2: 0.6001072644130586


### Random Forest Hiperparametre Optimizasyonu

In [5]:
# Random Forest Hiperparametre Ayarı
rf_model_optimized = RandomForestRegressor(n_estimators=200, max_depth=10, random_state=42)
rf_model_optimized.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_rf_optimized = rf_model_optimized.predict(test_data.drop(columns=['BAD']))

print("Optimized Random Forest MSE:", mean_squared_error(test_data['BAD'], y_pred_rf_optimized))
print("Optimized Random Forest MAE:", mean_absolute_error(test_data['BAD'], y_pred_rf_optimized))
print("Optimized Random Forest R2:", r2_score(test_data['BAD'], y_pred_rf_optimized))

Optimized Random Forest MSE: 0.0385163443576532
Optimized Random Forest MAE: 0.08643429118272444
Optimized Random Forest R2: 0.49510392065386943


### KNN

In [6]:
from sklearn.neighbors import KNeighborsRegressor

# KNN modeli oluşturma ve eğitme
knn_model = KNeighborsRegressor(n_neighbors=5)
knn_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_knn = knn_model.predict(test_data.drop(columns=['BAD']))

# Performans değerlendirmesi
print("KNN MSE:", mean_squared_error(test_data['BAD'], y_pred_knn))
print("KNN MAE:", mean_absolute_error(test_data['BAD'], y_pred_knn))
print("KNN R2:", r2_score(test_data['BAD'], y_pred_knn))

KNN MSE: 0.04410104011887073
KNN MAE: 0.05824665676077266
KNN R2: 0.4218962722852513


### KNN Hiperparametre Optimizasyonu

In [7]:
# KNN Hiperparametre Ayarı
knn_model_optimized = KNeighborsRegressor(n_neighbors=10, weights='distance')
knn_model_optimized.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_knn_optimized = knn_model_optimized.predict(test_data.drop(columns=['BAD']))

print("Optimized KNN MSE:", mean_squared_error(test_data['BAD'], y_pred_knn_optimized))
print("Optimized KNN MAE:", mean_absolute_error(test_data['BAD'], y_pred_knn_optimized))
print("Optimized KNN R2:", r2_score(test_data['BAD'], y_pred_knn_optimized))

Optimized KNN MSE: 0.045222122409229584
Optimized KNN MAE: 0.0669350798270925
Optimized KNN R2: 0.4072004317929513


### SVM

In [8]:
from sklearn.svm import SVR

# SVM modeli oluşturma ve eğitme
svm_model = SVR(kernel='rbf')
svm_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_svm = svm_model.predict(test_data.drop(columns=['BAD']))

# Performans değerlendirmesi
print("SVM MSE:", mean_squared_error(test_data['BAD'], y_pred_svm))
print("SVM MAE:", mean_absolute_error(test_data['BAD'], y_pred_svm))
print("SVM R2:", r2_score(test_data['BAD'], y_pred_svm))

SVM MSE: 0.036789804918952985
SVM MAE: 0.10175797677940297
SVM R2: 0.5177364681599775


### SVM Hiperparametre Optimizasyonu

In [9]:
# SVM Hiperparametre Ayarı
svm_model_optimized = SVR(kernel='rbf', C=10, epsilon=0.1)
svm_model_optimized.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_svm_optimized = svm_model_optimized.predict(test_data.drop(columns=['BAD']))

print("Optimized SVM MSE:", mean_squared_error(test_data['BAD'], y_pred_svm_optimized))
print("Optimized SVM MAE:", mean_absolute_error(test_data['BAD'], y_pred_svm_optimized))
print("Optimized SVM R2:", r2_score(test_data['BAD'], y_pred_svm_optimized))

Optimized SVM MSE: 0.02925000624378743
Optimized SVM MAE: 0.0963324395700462
Optimized SVM R2: 0.616572815524647


### Decision Tree

In [11]:
from sklearn.tree import DecisionTreeRegressor

# Decision Tree modeli oluşturma ve eğitme
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_dt = dt_model.predict(test_data.drop(columns=['BAD']))

# Performans değerlendirmesi
print("Decision Tree MSE:", mean_squared_error(test_data['BAD'], y_pred_dt))
print("Decision Tree MAE:", mean_absolute_error(test_data['BAD'], y_pred_dt))
print("Decision Tree R2:", r2_score(test_data['BAD'], y_pred_dt))

Decision Tree MSE: 0.06092124814264487
Decision Tree MAE: 0.06092124814264487
Decision Tree R2: 0.2014065755962029


### Decision Tree Hiperparametre Optimizasyonu

In [13]:
# Decision Tree Hiperparametre Ayarı
dt_model_optimized = DecisionTreeRegressor(max_depth=10, min_samples_split=5, random_state=42)
dt_model_optimized.fit(train_data.drop(columns=['BAD']), train_data['BAD'])
y_pred_dt_optimized = dt_model_optimized.predict(test_data.drop(columns=['BAD']))

print("Optimized Decision Tree MSE:", mean_squared_error(test_data['BAD'], y_pred_dt_optimized))
print("Optimized Decision Tree MAE:", mean_absolute_error(test_data['BAD'], y_pred_dt_optimized))
print("Optimized Decision Tree R2:", r2_score(test_data['BAD'], y_pred_dt_optimized))

Optimized Decision Tree MSE: 0.04851190413315176
Optimized Decision Tree MAE: 0.08730877590670366
Optimized Decision Tree R2: 0.3640759369320359


En İyi Model:
    
En Yüksek R²: Standartlaştırılmış SVM (R²: 0.616).
    

MSE: 0.029 (çok düşük, tahminler isabetli).
    
MAE: 0.096 (düşük, gerçek değerlere yakın).
    
Bu sonuç, modelin veriyi çok iyi açıkladığını gösteriyor.(SVM)


Alternatif Model: Random Forest (Normalize edilmiş veya standartlaştırılmış).
    
R²: 0.601 (standartlaştırılmış).
    
Performansı oldukça güçlü, ancak SVM'nin gerisinde.